src/language/extension-types.md

```diff
@@ -0,0 +1,315 @@
+
+```dart
+class C {
+  void m() {}
+}
+
+extension type E(C c) implements C {
+  @redeclare
+  void m() {}
+}
+```
+
+If you don't use the annotation, the compiler will choose the [more specific][]
+definition wherever the name is invoked, which might not be the behavior you expect.
+
+## Usage
+
+To use an extension type, create an instance the same as you would with a class:
+
+```dart
+extension type NumberE(int value) {
+  NumberE operator +(NumberE other) =>
+      NumberE(value + other.value);
+
+  NumberE get i => value;
+  bool isValid() => !value.isNegative;
+}
+
+void testE() { 
+  var num = NumberE(1);
+}
+```
+
+Then you can invoke members on the object as you would with a class object.
+Any attempt to treat the extension type object as an object
+of the representation type, or vice versa, will result in compile time errors
+(unless that behavior is explicitly defined for the extension type,
+or it is [transparent](#transparency)).
+For example:
+
+```dart
+void testE() { 
+  var num = NumberE(1);
+  int num2 = NumberE(2); // Error: Can't assign 'NumberE' to 'int'.
+  
+  num.isValid(); // Ok: Extension member invocation.
+  num.isNegative(); // Error: 'NumberE' does not define 'int' member 'isNegative'.
+  
+  var sum = num + num; // Ok: 'NumberE' defines '+'.
+  var diff = num - num; // Error: 'NumberE' does not define 'int' member '-'.
+  var diff2 = num.value - 2; // Ok: Can access representation object with reference.
+  var sum2 = num + 2; // Error: Can't assign 'int' to parameter type 'NumberE'. 
+  
+  List<NumberE> numbers = [
+    NumberE(1), 
+    num.i, // Ok: 'i' getter returns type 'NumberE'.
+    1, // Error: Can't assign 'int' element to list type 'NumberE'.
+    1 as NumberE, // Ok: Dynamic type cast recognizes representation type.
+  ];
+}
+```
+
+#### Transparency
+
+An extension type that [implements](#implements) its representation type
+is considered *transparent* because it gives the extension type access
+to *all* members of the representation type. Otherwise,
+only the members declared in the extension type body are available to it
+(considered a [*somewhat protected*](#type-considerations) extension type).
+
+```dart
+extension type NumberT(int value) 
+  implements int {
+  // Doesn't explicitly declare any members of 'int'.
+  NumberT get i => NumberT(value);
+}
+
+void main () {
+  var num = NumberT(1);
+  
+  // All OK: Representation type interface is available to extension type:
+  int numT = NumberT(2);
+  num.i - num; numT + num; num - 1; 2 + num;
+  
+  // Error: Extension type interface is not available to representation type:
+  numT.i;
+}
+```
+
+Transparency is a conceptual decision for the extension type creator to consider.
+A transparent extension type provides an *extended* interface to an existing type.
+You can invoke all the members of the representation type,
+plus any auxillary members you declare. The new interface is available
+to instances of the extension type.
```

Comment:
It's actually about expressions and their static type. Perhaps this will do:
```suggestion
for expressions of the extension type.
```
(OK, github refuses to specify a `suggestion` here, but it's intended to replace line 267.)